### PR TITLE
feat(security): path jails + schema bounds + crypto boundary

### PR DESCRIPTION
**Pull request by klodr**
## Summary

Security hardening foundation PR. Audit-driven fixes, no feature changes visible to existing users beyond the two new env-gated jails.

### Highest-impact fix

**Attachment path jail.** `createEmailWithNodemailer` accepted any absolute path the agent named. A crafted inbound email could instruct a prompt-injected agent to attach `~/.ssh/id_rsa`, `~/.gmail-mcp/credentials.json`, or `~/.claude.json` to an outgoing message. Paths are now canonicalized with `fs.realpathSync` and rejected unless they resolve inside `GMAIL_MCP_ATTACHMENT_DIR` (default `~/GmailAttachments`, mode `0o700`). Symlink escapes are blocked.

### Other security fixes

- **Download path jail** (`download_email` + `download_attachment`) under `GMAIL_MCP_DOWNLOAD_DIR` (default `~/GmailDownloads`). Replaces the prior `process.cwd()` fallback, which wrote attachments wherever the server happened to be launched.
- **Crypto multipart boundary** — `Math.random()` → `crypto.randomBytes(16).toString('hex')` in `createEmailMessage`.
- **Zod bounds** on `maxResults` (≤500 for search/threads, ≤100 for expanded threads), `batchSize` (≤100), `messageIds` length (≤1000).
- **Directories created by the MCP** are now always mode `0o700`.

### Housekeeping

- `engines.node`: `>=14.0.0` → `>=20.11` (matches CI matrix + the klodr MCP fleet)
- `@types/node`: `^20.10.5` → `^20.19.39`

## Test plan

- [x] `npm test` — **117/117 passing** (20 new tests in `src/hardening.test.ts`)
- [x] `npm run build` — clean
- [x] Covered scenarios: relative path, non-existent file, path outside jail, symlink escape, valid-in-jail path, boundary format + distinctness, schema bound edge cases

## What's NOT in this PR

Deliberately scoped foundation-only. These are next, on their own branches:

- Recipient pairing gate (`pair_recipient` tool + `~/.gmail-mcp/paired.json`) — analogue of `faxdrop_pair_number`
- Middleware (dry-run, audit log, per-tool dual-window rate limits)
- `SECURITY.md` + `ASSURANCE_CASE.md` + `CHANGELOG.md`

### CODE REVIEW

**Comment by klodr** [comment]
@coderabbitai review

**Comment by klodr** [comment]
@coderabbitai review

**Comment by klodr** [comment]
@coderabbitai review

**Comment by klodr** [comment]
@coderabbit fullreview

**Comment by klodr** [comment]
@coderabbit review

**Comment by klodr** [comment]
@coderabbit review

**Comment by klodr** [comment]
@coderabbit review

**Comment by klodr** [comment]
Latest commits address the 3 label-manager findings from review 4148366112:

- `cd0a71f` — 400 duplicate detection requires message corroboration; SystemLabelProtectionError sentinel prevents the system-label guard from being swallowed; getOrCreateLabel recovers from the findLabelByName→createLabel TOCTOU via rescan.

@coderabbitai review

**Comment by klodr** [comment]
@coderabbit review

**Comment by klodr** [comment]
@coderabbit review

**Comment by klodr** [comment]
All three findings from review 4148244347 are addressed in later commits:

- [3118172824] HTTP 400 no longer auto-classifies as duplicate — tightened in `cd0a71f` (src/label-manager.ts lines 60-66: 409 is unambiguous, 400 requires `.includes("already exists")`, no status falls back to message match alone).
- [3118172831] System-label guard no longer swallowed — `cd0a71f` introduces a `SystemLabelProtectionError` sentinel that the generic catch re-throws unchanged.
- [3118172843] `getOrCreateLabel` TOCTOU — `cd0a71f` wraps `createLabel` in a try/catch; when the racing duplicate error surfaces, we rescan with `findLabelByName` and return whatever actually landed.

@coderabbitai full review

**Comment by klodr** [comment]
@coderabbit fullreview

File: src/hardening.test.ts [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+  resetJailDirCache,
+} from "./utl.js";
+import {
+  SearchEmailsSchema,
+  BatchModifyEmailsSchema,
+  BatchDeleteEmailsSchema,
+  ListInboxThreadsSchema,
+  GetInboxWithThreadsSchema,
+} from "./tools.js";
+
```
[the remark this answers is not in the document]
Comment (klodr) [reply src/hardening.test.ts:31]: CRLF sanitization tests added in commit 49bc160 (6 regression tests covering subject, inReplyTo, references, from/cc/bcc injection attempts, null-byte stripping, legitimate-header passthrough). Tests 127/127 pass.

File: src/reply-all-helpers.test.ts [hunk 2]
```diff
@@ -1,207 +1,216 @@
...
+  it("is case insensitive when excluding authenticated user", () => {
+    const result = buildReplyAllRecipients(
+      "sender@example.com",
+      "ME@EXAMPLE.COM, other@example.com",
+      "",
+      myEmail,
+    );
+    expect(result.cc).toEqual(["other@example.com"]);
+  });
 });
```
[the remark this answers is not in the document]
Comment (klodr) [reply src/reply-all-helpers.test.ts:243]: Regression tests added in commit 38ebdaf (src/reply-all-helpers.test.ts lines 53-69). Two cases lock the quoted-comma parser: single \"Doe, John\" <...> + multiple quoted display names with commas. Tests 129/129 pass.

File: src/utl.test.ts [hunk 3]
```diff
@@ -10,83 +10,83 @@
...
-        expect(source).toContain('message-id');
-        expect(source).toContain('rfcMessageId');
-        expect(source).toContain('Message-ID: ${rfcMessageId}');
-    });
+  it("read_email returns Message-ID", () => {
+    const source = fs.readFileSync(path.join(srcDir, "index.ts"), "utf-8");
+    expect(source).toContain("message-id");
+    expect(source).toContain("rfcMessageId");
+    expect(source).toContain("Message-ID: ${rfcMessageId}");
+  });
```
[the remark this answers is not in the document]
Comment (klodr) [reply src/utl.test.ts:91]: Not addressed in this PR — the source-string style is a deliberate pattern in utl.test.ts (three 'Source verification' cases on lines 72-91) to assert that specific implementation patterns remain present after refactor. Switching to behavior-driven assertions would require mocking the Gmail API + nodemailer transport, which is outside the scope of the current security-hardening PR. Tracked as follow-up cleanup.

**Comment by klodr** [comment]
All prior review threads are now resolved (CRLF tests + reply-all regression + utl.test source-string explanation). Latest commits:
- `eca343f` — trim Scorecard trigger noise
- `ed0cfd4` — add code-audit as v1.0.0 blocker in roadmap
- `06895d2` — move Roadmap to ROADMAP.md
- `c98061c` — consolidate Silver/Gold wording
- `b4235c8` — promote ROADMAP sub-headings to H2

@coderabbitai full review

File: .github/workflows/verify-release.yml [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+            SIG_OK=$(printf '%s' "$META" | jq -r '.dist.signatures | length // 0')
+            if [ "$VER_OK" = "$VERSION" ] && [ -n "$ATT_OK" ] && [ "${SIG_OK:-0}" -ge 1 ]; then
+              echo "npm OK: ${PKG}@${VERSION} (attestations: $ATT_OK, signatures: $SIG_OK)"
+              exit 0
+            fi
+            echo "Waiting for ${PKG}@${VERSION} (version=$VER_OK, att=${ATT_OK:-<missing>}, sigs=${SIG_OK:-0}) — $i/10"
+            sleep 10
+          done
+          echo "Timeout waiting for ${PKG}@${VERSION} (version+attestations+signatures)" >&2
+          exit 1
```
[the remark this answers is not in the document]
Comment (klodr) [reply .github/workflows/verify-release.yml]: Fixed in `6db28ab`. Extended the npm-metadata polling window from 10×10s = 100s to 15 attempts with 10→20s backoff (~5 min total) so npm attestation/signature propagation lag no longer flags as timeout.

File: package.json [hunk 5]
```diff
@@ -1,69 +1,89 @@
...
+    "mcp-evals": "^2.0.1",
+    "prettier": "^3.8.3",
+    "tsup": "^8.5.1",
+    "typescript": "^6.0.3",
+    "typescript-eslint": "^8.58.2",
+    "vitest": "^4.1.4"
+  },
+  "overrides": {
+    "test-exclude": "^8.0.0"
   }
```
[the remark this answers is not in the document]
Comment (klodr) [reply package.json]: Fixed in `6db28ab`. Added a `//` JSON-comment in the overrides block explaining the test-exclude@^8.0.0 pin (vitest 4.x + coverage-v8 compatibility; 9.x's ESM rewrite breaks the istanbul globs).

File: src/label-manager.ts [hunk 6]
```diff
@@ -61,32 +71,37 @@ export async function createLabel(gmail: any, labelName: string, options: {
...
+    });
+
+    return response.data;
+  } catch (err: unknown) {
+    const error = asGmailApiError(err);
+    if (error.code === 404) {
+      throw new Error(`Label with ID "${labelId}" not found.`, { cause: err });
     }
+
+    throw new Error(`Failed to update label: ${error.message}`, { cause: err });
```
[the remark this answers is not in the document]
Comment (klodr) [reply src/label-manager.ts:117]: Fixed in `6db28ab`. Dropped the preflight `labels.get()` call — `labels.update()` is the source of truth and the catch already translates 404 cleanly.

File: src/label-manager.ts [hunk 7]
```diff
@@ -183,21 +214,41 @@ export async function findLabelByName(gmail: any, labelName: string) {
...
-    } catch (error: any) {
-        throw new Error(`Failed to get or create label: ${error.message}`);
+      return await createLabel(gmail, labelName, options);
+    } catch (createErr: unknown) {
+      const createMsg = createErr instanceof Error ? createErr.message : String(createErr);
+      if (createMsg.includes("already exists")) {
+        const racedLabel = await findLabelByName(gmail, labelName);
+        if (racedLabel) return racedLabel;
+      }
+      throw createErr;
```
[the remark this answers is not in the document]
Comment (klodr) [reply src/label-manager.ts:247]: Fixed in `6db28ab`. Introduced a `DuplicateLabelError` sentinel class thrown by `createLabel()`. The TOCTOU recovery in `getOrCreateLabel()` now checks `instanceof DuplicateLabelError` instead of grep-ing the user-facing message text, so a future copy change cannot silently disable the recovery.

**Comment by klodr** [comment]
Follow-up on the test-exclude override discussion — commit `8399296` supersedes `6db28ab`. The `//` JSON-comment I added first broke `npm ci` (npm treats every key under `overrides` as a package name and rejected `//` with 'Override without name'). Moving the rationale to an adjacent `_overridesRationale` key exposed the fact that `test-exclude` isn't even a transitive dep on the current vitest 4.1.4 tree (`npm ls test-exclude` → empty). Dropped the override outright.

**Comment by klodr** [comment]
@coderabbit review

File: src/reply-all-helpers.test.ts [hunk 8]
```diff
@@ -1,207 +1,234 @@
...
+    ]);
+  });
+
+  it("handles multiple quoted display names with commas", () => {
+    expect(
+      parseEmailAddresses(
+        '"Smith, Alice" <alice@example.com>, "Jones, Bob" <bob@example.com>, "Brown, Carol" <carol@example.com>',
+      ),
+    ).toEqual(["alice@example.com", "bob@example.com", "carol@example.com"]);
+  });
```
[the remark this answers is not in the document]
Comment (klodr) [reply src/reply-all-helpers.test.ts:69]: Added in `9f2fc7e` (src/reply-all-helpers.test.ts lines 71-78). The parser already handled this via the `headerValue[i-1] !== "\\"` guard; the new test locks the behaviour. 130/130 tests pass.
